src/t_hash.c

```diff
@@ -297,25 +300,399 @@ static void hashDictWithExpireOnRelease(dict *d) {
+        prevExpire = (uint64_t) expireTime;
+    }
+
+    if (prevExpire == EB_EXPIRE_TIME_INVALID) {
+        if (ex->expireSetCond & (HFE_XX | HFE_LT | HFE_GT))
+            return HSETEX_NO_CONDITION_MET;
+    } else {
+        if (((ex->expireSetCond == HFE_GT) && (prevExpire >= expireAt)) ||
+            ((ex->expireSetCond == HFE_LT) && (prevExpire <= expireAt)) ||
+            (ex->expireSetCond == HFE_NX) )
+            return HSETEX_NO_CONDITION_MET;
+
+        /* Track of minimum expiration time (only later update global HFE DS) */
+        if (ex->minExpireFields > prevExpire)
+            ex->minExpireFields = prevExpire;
+    }
+
+    /* if expiration time is in the past */
+    if (unlikely(checkAlreadyExpired(expireAt))) {
+        hashTypeDelete(ex->hashObj, field);
+        ex->fieldDeleted++;
+        return HSETEX_DELETED;
+    }
+
+    if (ex->minExpireFields > expireAt)
+        ex->minExpireFields = expireAt;
+
+    listpackExUpdateExpiry(ex->hashObj, field, fptr, vptr, expireAt);
+    ex->fieldUpdated++;
+    return HSETEX_OK;
+}
+
+/* Returns 1 if expired */
+int hashTypeIsExpired(const robj *o, uint64_t expireAt) {
+    if (o->encoding == OBJ_ENCODING_LISTPACK_EX) {
+        if (expireAt == HASH_LP_NO_TTL)
+            return 0;
+    } else if (o->encoding == OBJ_ENCODING_HT) {
+        if (expireAt == EB_EXPIRE_TIME_INVALID)
+            return 0;
+    } else {
+        serverPanic("Unknown encoding: %d", o->encoding);
+    }
+
+    return (mstime_t) expireAt < commandTimeSnapshot();
+}
+
+/* Returns listpack pointer of the object. */
+unsigned char *hashTypeListpackGetLp(robj *o) {
+    if (o->encoding == OBJ_ENCODING_LISTPACK)
+        return o->ptr;
+    else if (o->encoding == OBJ_ENCODING_LISTPACK_EX)
+        return ((listpackEx*)o->ptr)->lp;
+
+    serverPanic("Unknown encoding: %d", o->encoding);
+}
+
+size_t hashTypeListpackMemUsage(robj *o) {
+    if (o->encoding == OBJ_ENCODING_LISTPACK) {
+        return sizeof(*o) + zmalloc_size(o->ptr);
+    } else if (o->encoding == OBJ_ENCODING_LISTPACK_EX) {
+        listpackEx *lpt = o->ptr;
+        return sizeof(*o) + zmalloc_size(lpt) + zmalloc_size(lpt->lp);
+    }
+    serverPanic("Unknown encoding: %d", o->encoding);
+}
+
 /*-----------------------------------------------------------------------------
  * Hash type API
  *----------------------------------------------------------------------------*/
 
 /* Check the length of a number of objects to see if we need to convert a
  * listpack to a real hash. Note that we only check string encoded objects
  * as their string length can be queried in constant time. */
-void hashTypeTryConversion(robj *o, robj **argv, int start, int end) {
+void hashTypeTryConversion(redisDb *db, robj *o, robj **argv, int start, int end) {
     int i;
     size_t sum = 0;
 
-    if (o->encoding != OBJ_ENCODING_LISTPACK) return;
+    if (o->encoding != OBJ_ENCODING_LISTPACK &&
+        o->encoding != OBJ_ENCODING_LISTPACK_EX)
```

Comment:
88 columns is not too long.
```suggestion
    if (o->encoding != OBJ_ENCODING_LISTPACK && o->encoding != OBJ_ENCODING_LISTPACK_EX)
```